app/lib/brexit_checker/results_audiences.rb

```diff
@@ -1,65 +1,68 @@
+    {
+      actions: business_actions,
+      criteria: business_actions.flat_map(&:all_criteria).uniq & criteria,
+    }
+  end
 
-        group_keys =
-          multiple_grouping_criteria?(action) ? selected_group(action, selected_criteria) : action.grouping_criteria
-        group_keys.each do |key|
-          group = all_groups.select { |g| g.key == key }.first
-          next if grouped_actions.select { |actions| actions[:group] == group }.any?
+  def populate_citizen_groups
+    return [] if citizen_actions.blank? || criteria.blank?
 
-          grouped_actions << {
-            group: group,
-            actions: actions_for_group(audience_actions, group),
-            criteria: criteria_for_group(audience_actions, group, selected_criteria),
-          }
-          grouped_actions.sort_by! { |actions| [(actions[:group].priority * -1), actions[:group].key] }
-        end
+    citizen_actions.each_with_object([]) do |action, grouped_actions|
+      next if action.grouping_criteria.empty?
+
+      group_keys =
+        action.multiple_grouping_criteria? ? selected_group(action) : action.grouping_criteria
+      group_keys.each do |key|
+        group = all_groups.select { |g| g.key == key }.first
+        next if grouped_actions.select { |actions| actions[:group] == group }.any?
+
+        grouped_actions << {
+          group: group,
+          actions: citizen_actions & group.actions,
+          criteria: criteria_for_group(group),
+        }
+        grouped_actions.sort_by! { |actions| [(actions[:group].priority * -1), actions[:group].key] }
       end
     end
+  end
 
-    def selected_group(action, selected_criteria)
-      criteria_keys = selected_criteria.map(&:key)
-      action.grouping_criteria.select { |group_key| criteria_keys.include? group_key }
-    end
+private
 
-    def all_groups
-      @all_groups ||= BrexitChecker::Group.load_all
-    end
+  def selected_group(action)
+    action.grouping_criteria.select { |group_key| criteria_keys.include? group_key }
+  end
 
-    def actions_for_group(audience_actions, group)
-      audience_actions & group.actions
-    end
+  def all_groups
+    @all_groups ||= BrexitChecker::Group.load_all
+  end
 
-    def criteria_for_group(audience_actions, group, selected_criteria)
-      actions = audience_actions & group.actions
-      filtered_criteria =
-        actions.map do |action|
-          if multiple_grouping_criteria?(action)
-            rogue_criterion = (action.grouping_criteria - [group.key]).first
-            (action.all_criteria & selected_criteria).reject { |criteria| criteria.key == rogue_criterion }
-          else
-            action.all_criteria & selected_criteria
-          end
-        end
-      filtered_criteria.flatten.uniq
-    end
+  def criteria_keys
+    @criteria_keys ||= criteria.map(&:key)
+  end
 
-    def multiple_grouping_criteria?(action)
-      action.grouping_criteria.count > 1
-    end
+  def criteria_for_group(group)
```

Comment:
wonder if it's worth naming this as being specific to citizens?

Reply:
I've refactored so that the actions are passed in.  If we add groupings for business actions we can reuse the method, in theory!